Initial migration of site to 11ty (#5483)
dart-lang/site-www: README.md

[![Build Status SVG][]][Repo on GitHub Actions]
[![OpenSSF Scorecard SVG][]][Scorecard Results]
[![first-timers SVG][]][first-timers]

The https://dart.dev site, built with [Jekyll][] and hosted on [Firebase][].
The documentation site for the [Dart programming language](https://dart.dev), 
built with [Eleventy][] and hosted on [Firebase][].

[We welcome contributions](CONTRIBUTING.md),

> and we're [first-timer friendly][first-timers]!

Author: We are! But when I was updating the README I realized the label isn't there anymore, so I removed this link for now to avoid any confusion.

I have follow-up work to introduce more contributing docs for the site infra, syntax, and writing practices. Once that's done and this setup is finalized, I'll bring something back here so it's actually easy to get going as a "first-timer".

check out the [Contributing guidelines](CONTRIBUTING.md).

## Getting started

Start by looking for an [issue](https://github.com/dart-lang/site-www/issues)
that catches your interest, or create an issue with your proposed change.
Consider adding a comment to let everyone know that you're working on it, and
feel free to ask any questions you have on the same issue.

To update this site, fork the repo, make your changes,
and generate a pull request.
For simple changes (such as style and typo fixes),
you probably don't need to build this site.
Often you can make changes using the GitHub UI.

> We can stage the changes automatically in your pull request.

Reviewer: Great call out!

Author: It was just an adjustment of what was there, but thanks :P

- While the new setup is easier, I still want to discourage developers looking to quickly contribute.

> [!IMPORTANT]  
> If you are cloning this repository locally,
> follow the below instruction on cloning with its submodule.

If your change involves code samples, adds/removes pages, or affects navigation,
do consider building and test your work before submitting.

If you want or need to build the site, follow the steps below.

## Build the site

For changes beyond simple text and CSS tweaks,
we recommend running the site locally to
enable an edit-refresh cycle. 

### Get the prerequisites

Install the following tools, if you don't have them already:

- **Dart**
  The latest stable release of Dart is required to build the site
  and run its tooling. This can be the Dart included in the Flutter SDK.
  If you don't have Dart or need to update, follow the
  instructions at [Get the Dart SDK][].

- **Node.js**

>   The latest stable LTS release of Node.js is required to build the site.

Author: I've never used nvm, but this seems like an easy win for those that do. Added! Is it popular enough that I should add a mention to the README?

Thanks :)

>   The latest stable LTS release of Node.js is required to build the site.

Reviewer 2: @parlough : Indeed it is.

  then run this command from the repo root:

  ```terminal
  git submodule update --init --recursive
  ```

> [!NOTE]
> At any time during development
> you can use the `git submodule` command to refresh submodules:
>
> ```terminal
> git pull && git submodule update --init --recursive
> ```

## Set up your local environment and serve changes

1. _Optional:_ After cloning the repo and its submodules,
   create a branch for your changes:

   ```terminal
   git checkout -b <BRANCH_NAME>
   ```

2. From the root directory of the repository,
   fetch the site's Dart dependencies.

   ```terminal
   dart pub get
   ```

3. From the root directory of the repository,
   enable `corepack` to set up `pnpm`, then
   fetch the site's npm dependencies.
   If you already have `pnpm` installed,
   you can skip the `corepack` commands.

   ```terminal
   corepack enable
   corepack install
   pnpm install
   ```

4. From the root directory, serve the site locally.

   ```terminal
   dart run dart_site serve
   ```

   This command generates and serves the site on a
   local port that's printed to your terminal.

> 5. View your changes in the browser by navigating to `http://localhost:4000`.

Author: Good idea, thanks!

>    The port might be different if `4000` is taken.
>    If you need to view the generated fields,
>    they can be found in the `_site` directory.

Reviewer 3: ```suggestion
   > [!NOTE]
   > The port might be different if `4000` is taken. For generated site files,
   > see the `_site` directory.
```

> you can use the `git submodule` command to refresh submodules:
>
> ```terminal
> git pull && git submodule update --init --recursive
> ```

## Set up your local environment and serve changes

1. _Optional:_ After cloning the repo and its submodules,
   create a branch for your changes:

   ```terminal
   git checkout -b <BRANCH_NAME>
   ```

2. From the root directory of the repository,
   fetch the site's Dart dependencies.

   ```terminal
   dart pub get
   ```

3. From the root directory of the repository,
   enable `corepack` to set up `pnpm`, then
   fetch the site's npm dependencies.
   If you already have `pnpm` installed,
   you can skip the `corepack` commands.

   ```terminal
   corepack enable
   corepack install
   pnpm install
   ```

4. From the root directory, serve the site locally.

   ```terminal
   dart run dart_site serve
   ```

   This command generates and serves the site on a
   local port that's printed to your terminal.

5. View your changes in the browser by navigating to <http://localhost:4000>.

   > [!NOTE]
   > The port might be different if `4000` is taken.
   
   To instead view the source of generated site files,
   check the `_site` directory.

> 

Reviewer: is "fields" supposed to be "files" here?

Author: It is already "files", but yes :)

6. Make your changes to the local repo.

   The site should automatically rebuild on most changes, but if

>    something doesn't update, exit the process and rerun the command.

Reviewer: Are there "known issues" here that we can flag for folks?

Reviewer: alternatively, an issue link or something

Author: I'd like to see which issues come up naturally, but I'll leave a comment to open an issue. Good point!

   Improvements to this functionality are planned.

7. Commit your changes to the branch and submit your PR.

   If your change is large, or you'd like to test it,
   consider [validating your changes](#validate-your-changes) and
   [deploying to a staging site](#deploy-to-a-staging-site).

> [!TIP]
> To find additional commands that you can run,
> run `dart run dart_site --help` from the repository's root directory.

## Validate your changes

### Check documentation and example code

If you've made changes to the code in the `/examples` or `/tool` directories,
commit your work, then run the following command to
verify it is up to date and matches the site standards.

```terminal
dart run dart_site check-all
```

If this script reports any errors or warnings,
then address those issues and rerun the command.

> If you have any issues, reach out for help.

Author: I'm not sure how I forgot to include something like this... "reach out for help" is useless :P

I added some more information as well as a link to the Flutter contributors Discord.

Thanks!!!

Submitted pull requests can be automatically staged
by a site maintainer.
If you'd like to stage the site yourself though,
you can build a full version and upload it to Firebase.

1. If you don't already have a Firebase project,

   - Navigate to the [Firebase Console](https://console.firebase.google.com)
     and create your own Firebase project (for example, `dart-dev-staging`).

   - Head back to your local terminal and verify that you are logged in.

     ```terminal
     firebase login
     ```

   - Ensure that your project exists and activate that project:

     ```terminal
     firebase projects:list
     firebase use <your-project>
     ```

2. From the root directory of the repository, build the site:

   ```terminal
   dart run dart_site build
   ```

   This will build the site and copy it to your local `_site` directory.
   If that directory previously existed, it will be replaced.

3. Deploy to your activated Firebase project's default hosting site:

   ```terminal
   firebase deploy --only hosting
   ```

4. Navigate to your PR on GitHub and include the link of the staged version.
   Do consider adding a reference to the commit you staged,

>    so that reviewers know if any further changes have been made.

Reviewer: doesn't our staging GitHub build handle this?

Author: Yes it does. This entire section is for if you'd like to stage on your own or you need to for some reason.

Reviewer: Can we perhaps add an [Optional] to the title intro section at the top title? Want to flag this is not required and folks should only do it if they want to for their own purposes.
